### PR TITLE
feat: add support to got http lib [RD-3047]

### PR DESCRIPTION
**Pull request by doriaviram**

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -233,8 +233,6 @@ workflows:
         filters:
           branches:
             ignore: master
-        requires:
-        - test
```
Comment (doriaviram) [inline .circleci/config.yml:237]: Ignore this part, not going to push this.
It's just, for now, I want the integration will run. before implementing all UT

File: src/hooks/http.js [hunk 2]
```diff
@@ -68,14 +68,35 @@ export const parseHttpRequestOptions = (options = {}, url) => {
...
+    body: '', // XXX Filled by the httpRequestEndWrapper or httpRequestEmitWrapper ( / Write)
     method,
     headers: lowerCaseObjectKeys(headers),
     protocol,
     sendTime,
   };
 };
 
+export const httpRequestEmitWrapper = requestData => emitFunc =>
+  function(eventName, arg) {
```
Comment (nirLumigo) [inline src/hooks/http.js:80]: Why isn't it an arrow function like all the others?

File: src/hooks/http.js [hunk 3]
```diff
@@ -68,14 +68,35 @@ export const parseHttpRequestOptions = (options = {}, url) => {
...
+      () => {
+        if (arg && arg._httpMessage && arg._httpMessage._hasBody) {
+          const httpMessage = arg._httpMessage;
+          let lines = [];
+          if (httpMessage.hasOwnProperty('outputData')) {
+            lines = httpMessage.outputData[0].data.split('\n');
+          } else if (httpMessage.hasOwnProperty('output')) {
+            lines = httpMessage.output[0].split('\n');
+          }
+          requestData.body += lines[lines.length - 1];
```
Comment (nirLumigo) [inline src/hooks/http.js:91]: Are we adding lines to the body of the request?

File: src/hooks/http.js [hunk 4]
```diff
@@ -159,16 +182,16 @@ export const httpRequestArguments = args => {
   let options = undefined;
   let callback = undefined;
 
-  if (typeof args[0] === 'string') {
+  if (typeof args[0] === 'string' || args[0] instanceof URL) {
```
Comment (nirLumigo) [inline src/hooks/http.js:185]: Why `instanceof` and not `typeof`?

File: src/hooks/http.test.js [hunk 5]
```diff
@@ -296,7 +389,39 @@ describe('http hook', () => {
...
+    const options = { a: 'b' };
+    const expected1 = {
+      url,
+      options,
+      callback: undefined,
+    };
+    expect(httpHook.httpRequestArguments([url, options])).toEqual(expected1);
+  });
+
+  test('httpRequestArguments -> http(objectUrl, callback)', () => {
```
Comment (nirLumigo) [inline src/hooks/http.test.js:413]: Shouldn't there be also a test with `http(callback)`?

File: src/hooks/http.js [hunk 6]
```diff
@@ -68,14 +68,35 @@ export const parseHttpRequestOptions = (options = {}, url) => {
...
 };
 
+export const httpRequestEmitWrapper = requestData => emitFunc =>
+  function(eventName, arg) {
+    safeExecute(
+      () => {
+        if (arg && arg._httpMessage && arg._httpMessage._hasBody) {
+          const httpMessage = arg._httpMessage;
+          let lines = [];
+          if (httpMessage.hasOwnProperty('outputData')) {
```
Comment (saartochner) [inline src/hooks/http.js:86]: validate that you can run `httpMessage.outputData[0].data.split`, i.e.  that data is string, etc.

File: src/hooks/http.js (hunk 3 above)
Comment (saartochner) [inline src/hooks/http.js:91]: why do you take only the last line?
Don't you want to take all the lines to the body?

File: src/hooks/http.js [hunk 7]
```diff
@@ -267,12 +292,20 @@ export const httpRequestWrapper = originalRequestFn =>
         logger.warn('end wrap error', e.message);
       }
 
+      if (!isRequestToAwsService) {
```
Comment (saartochner) [inline src/hooks/http.js:295]: huge smell-
Can you share what happened when you did it to aws requests?

Comment (saartochner) [reply src/hooks/http.js:295]: If it's destructive, then we might destruct customer's requests...

File: src/hooks/http.js [hunk 8]
```diff
@@ -267,12 +292,20 @@ export const httpRequestWrapper = originalRequestFn =>
...
+        try {
+          const emitWrapper = httpRequestEmitWrapper(requestData);
+          shimmer.wrap(clientRequest, 'emit', emitWrapper);
```
Comment (saartochner) [inline src/hooks/http.js:298]: you're adding another shimmering - can you stress test that?

File: src/hooks/http.js (hunk 3 above)
Comment (saartochner) [reply src/hooks/http.js:91]: Moreover, if lines is empty, then you mess `requestData.body` with `+=undefined` which cast it to string 😂

**Comment by saartochner** [comment]
I'm missing an "e2e" test that uses the `emit` functionality and checks that we collected the right data

File: src/hooks/http.js (hunk 6 above)
Comment (doriaviram) [reply src/hooks/http.js:86]: It's wrapped in `safeExecute`

File: src/hooks/http.js (hunk 8 above)
Comment (doriaviram) [reply src/hooks/http.js]: Yes, I'm adding in integration-test

File: src/hooks/http.js (hunk 3 above)
Comment (doriaviram) [reply src/hooks/http.js]: Fixed :)

File: src/hooks/http.test.js (hunk 5 above)
Comment (doriaviram) [reply src/hooks/http.test.js:413]: Didn't find this case in the docs

File: src/index.test.js [hunk 9]
```diff
@@ -15,6 +15,7 @@ describe('index', () => {
     Object.keys(spies).map(x => spies[x].mockClear());
   });
 
+  jest.setTimeout(15000);
```
Comment (saartochner) [inline src/index.test.js:18]: put it inside the relevant test (and not on the class level)

File: src/utils.js [hunk 10]
```diff
@@ -497,3 +497,17 @@ export const md5Hash = item => {
   md5sum.update(JSON.stringify(item));
   return md5sum.digest('hex');
 };
+
+export const isEncodingType = encodingType =>
+  !!(
+    encodingType &&
+    typeof encodingType === 'string' &&
+    ['ascii', 'utf8', 'utf16le', 'ucs2', 'base64', 'binary', 'hex'].includes(
```
Comment (saartochner) [inline src/utils.js:505]: there's also `utf-8`, `utf-16-le` etc.:
https://stackoverflow.com/questions/14551608/list-of-encodings-that-node-js-supports
maybe you can import this list from somewhere else?

File: src/utils.js [hunk 11]
```diff
@@ -497,3 +497,17 @@ export const md5Hash = item => {
...
+      encodingType
+    )
+  );
+
+export const isEmptyString = str => !!(!str || str.length === 0);
```
Comment (saartochner) [inline src/utils.js:510]: add `typeof` to avoid `isEmptyString([])`

File: src/utils.js [hunk 12]
```diff
@@ -497,3 +497,17 @@ export const md5Hash = item => {
...
+
+export const isValidHttpRequestBody = reqBody =>
+  !!(reqBody && (typeof reqBody === 'string' || reqBody instanceof Buffer));
```
Comment (saartochner) [inline src/utils.js:513]: empty string is not valid?

File: src/hooks/http.test.js [hunk 13]
```diff
@@ -130,6 +130,172 @@ describe('http hook', () => {
     expect(requestData).toEqual({ body: '' });
   });
 
+  test('httpRequestWriteWrapper -> simple flow -> write(str)', dontExitWithoutDone => {
+    const requestData = {
+      body: '',
+    };
+    const wrapper = httpHook.httpRequestWriteWrapper(requestData);
```
Comment (saartochner) [inline src/hooks/http.test.js:137]: can you lower this line to where you use it?
It'll help the reader...

File: src/index.test.js (hunk 9 above)
Comment (doriaviram) [reply src/index.test.js:18]: I will move this, but Its doesn't matter in fact

File: src/hooks/http.js [hunk 14]
```diff
@@ -146,7 +169,7 @@ export const wrappedHttpResponseCallback = (
 
 export const httpRequestEndWrapper = requestData => originalEndFn =>
   function(...args) {
-    if (args[0] && (typeof args[0] === 'string' || args[0] instanceof Buffer)) {
```
[the remark this answers is not in the document]
Comment (doriaviram) [reply src/hooks/http.js]: 👍

File: src/utils.js (hunk 10 above)
Comment (doriaviram) [reply src/utils.js:505]: Its buffer encodings found this list here
https://stackoverflow.com/questions/34476754/what-encodings-does-buffer-tostring-support

File: src/utils.js (hunk 11 above)
Comment (doriaviram) [reply src/utils.js:510]: 👍

File: src/utils.js (hunk 12 above)
Comment (doriaviram) [reply src/utils.js:513]: No, it's mean its unfilled

File: src/hooks/http.test.js (hunk 13 above)
Comment (doriaviram) [reply src/hooks/http.test.js:137]: 👍

**Pull request merged**